common/all_llvm_targets.cpp

```diff
@@ -0,0 +1,20 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "common/init_llvm.h"
+#include "llvm/Support/TargetSelect.h"
+
+namespace Carbon {
+
+static auto InitLLVMTargets() -> void {
+  llvm::InitializeAllTargetInfos();
+  llvm::InitializeAllTargets();
+  llvm::InitializeAllTargetMCs();
+  llvm::InitializeAllAsmParsers();
+  llvm::InitializeAllAsmPrinters();
+}
+
+char InitLLVM::RegisterTargets = (InitializeTargets = &InitLLVMTargets, 0);
```

Comment:
This is probably just my naivette, but why:

- Store `RegisterTargets` as a member of `InitLLVM` versus something local to this file?
- Initialize using a comma expression instead of something like `auto Register = InitializeTargets = &InitLLVMTargets;`?

Mild thought, had you considered something like two different targets, one for init-with-all-targets, one for init-without-all-targets, resulting in something like a cpp file that only contains `auto (*InitLLVM::InitializeTargets)() -> void = nullptr;` versus `auto (*InitLLVM::InitializeTargets)() -> void = &InitLLVMTargets;` here? i.e., creating `:init_llvm_with_targets` and `:init_llvm_without_targets`, never directly depending on `:init_llvm`.

Reply:
> This is probably just my naivette, but why:
> 
> * Store `RegisterTargets` as a member of `InitLLVM` versus something local to this file?

So that it has access to the private member `InitLLVM::InitializeTargets`. That seems like it at least deserves a comment :)

> * Initialize using a comma expression instead of something like `auto Register = InitializeTargets = &InitLLVMTargets;`?

Just a trivial thing: that would make the type of `Register` be a function pointer, which would be a bit uglier, take up a bit more storage, and create a temptation to use `InitLLVM::RegisterTargets` instead of `InitLLVM::InitializeTargets` (given that they would store the same value).

> Mild thought, had you considered something like two different targets, one for init-with-all-targets, one for init-without-all-targets, resulting in something like a cpp file that only contains `auto (*InitLLVM::InitializeTargets)() -> void = nullptr;` versus `auto (*InitLLVM::InitializeTargets)() -> void = &InitLLVMTargets;` here? i.e., creating `:init_llvm_with_targets` and `:init_llvm_without_targets`, never directly depending on `:init_llvm`.

I did think about that. The consequence would be that every binary and test (transitively) using `//common:init_llvm` would need to link against one or the other, or get a link-time error. I'm fine with that, but it's a tradeoff -- do we want to burden (essentially) every `cc_binary` and `cc_test` rule with a dependency that says "don't link in the LLVM targets"? Maybe we should; it'll be clearer and more explicit that way, if more verbose.

I'll give that a go and see just how many extra deps I need to add. I think that would also address your other comments here (we can get rid of this registration machinery and function pointer shenanigans entirely, and just have two definitions of an `InitLLVM::InitializeTargets()` function).

Reply:
Added comments.